File: project/6.08_コンストラクタの修飾子/java/java_ex6_08_1/src/java_ex6_08_1/Ex6_08_1.java
package java_ex6_08_1;

class Person {
	String name;
	int age;
	
	public Person( String name, int age ) {
		this();					// privateは同じクラスから呼出し可能
		this.name = name;
		this.age = age;
	}
	
	private Person() {
		
	}
	
	protected Person( String name ) {
		this.name = name;
	}
}

class Child extends Person{
	Child( String name ) {
		// スーパークラスのコンストラクタにprivateがあるため、
		// どのコンストラクタを呼び出すのか、明示的に指示する必要がある
//		super();			// privateはサブクラスから呼出しできない
		super( name );		// protectedはサブクラスから呼出し可能
		super.age = 10;
	}
}

public class Ex6_08_1 {

	public static void main(String[] args) {
		Person p1 = new Person( "John", 30);
		System.out.println( p1.name );
		System.out.println( p1.age );
		
//		Person p2 = new Person();			// privateなクラスは呼び出せない
		Person p2 = new Person( "Keroichi" );
		System.out.println( p2.name );
		
		p2 = new Child( "Son" );
		System.out.println( p2.name );
		System.out.println( p2.age );
	}

}

File: project/6.08_コンストラクタの修飾子/py/py_ex6_08_1/src/ex6_08_1.py
'''
Created on 2023/05/11

@author: Keroichi.T
'''

class Person( object ):
    # クラス変数
    name = "Nothing"
    age = 0

    # コンストラクタへの修飾子はない    
    def __init__(self, name, age):
        self.name = name
        self.age = age
    
class Child( Person ):
    pass
    # def __init__(self, name, age):
    #     super().__init__(name, age)
    #     self.gender = "Man"
    

def main():
    print( Person.name )
    print( Person.age )
    
    p1 = Person( "John", 30 )
    print( p1.name )
    print( p1.age )
    
    print( Person.name )
    print( Person.age )

    p2 = Child( "Son", 10 )
    print( p2.name )
    print( p2.age )

if __name__ == '__main__':
    main()